add data validation step (#95)
epiverse-trace/tutorials-early: episodes/clean-data.Rmd

)
utils::head(data, 7)
```
The resulting linelist object resembles a dataframe but offers richer features 
and functionalities. Packages that are linelist-aware can leverage these 
features. For example, you can extract a dataframe of only the tagged columns 
using the tags_df() function, as shown below:
```{r, warning=FALSE}
head(linelist::tags_df(data), 5)

> ```

Reviewer: I would move this to the end. This will help to keep the following sequence of steps: tag, safeguarding, validate, and then use tagged-only columns downstream.

Author: OK. Will move it down.

Safeguarding is implicitly built into the linelist objects. If you try to delete any of the tagged 
columns, you will receive an error or warning message, as shown in the example below.

```{r, warning=TRUE}

> new_df <- data |> dplyr::select(linelist::has_tag(c("id", "age")))

Author: Good. Will add it

> new_df <- data |> dplyr::select(linelist::has_tag(c("id", "age")))

Author: accept it

```

> 

Reviewer: here I would add a challenge to test the complementary message of error and invite learners to identify how it changes the output and workflow downstream

```r
linelist::lost_tags_action(action = "error")
```

Author: Will add  a challenge

The {linelist} package supplies tags for the commen epidemiological variables 
and specify them the appropriate data types. You can view this by running the 
following command:
```{r, warning=FALSE}
linelist::tags_types()

> ```

Reviewer: I would move this as a hint in a challenge right after using `linelist::make_linelist()`. The output is long and more relevant when the learner is challenged to tag a different set of variables

Reviewer: and emphasize using a combination `dplyr::glimpse` with  `linelist::tags_names` to write new tags

Author: Ok. Will fix it.

To ensure that all tagged variables are standardized and have the correct data 
types, use the validate_tags() and validate_types() functions, respectively, as 
shown in the example below:

```r
linelist::validate_tags(data,
  allow_extra = FALSE
)
linelist::validate_types(data,
  ref_types = tags_types()
)
```
If your dataset contains a `non-default` tag, set the argument 

> `allow_extra = TRUE` when creating the linelist object.

Reviewer: instead of showing both functions, we can only show one `linelist::validate_linelist()` which embeds them ([check the function definition](https://github.com/epiverse-trace/linelist/blob/main/R/validate_linelist.R#L62-L70)).

We can make a subset from the raw data and provide an accidental tagging of a non-completely cleaned dataset, similar to:

```r
linelist::make_linelist(
  x = cleaned_data,
  id = "case_id",
  age = "gender" # accidental tagging
) %>% 
  linelist::validate_linelist()
```

possibly similar to a challenge.

> `allow_extra = TRUE` when creating the linelist object.

Reviewer: As a last resource, I think it appropriate to show two additional concepts, possibly in a hidden callout after a challenge or similar:
- how to add extra tags? with the `allow_extra = TRUE` argument
- how to validate these extra tags? by defining their specific `tags_types()`.

``` r
library(outbreaks)
library(tidyverse)
library(linelist)

# cleaned data
ebola <- outbreaks::ebola_sim_clean %>% 
  pluck("linelist") %>% 
  as_tibble()

# tag data
ebola_tag <- ebola %>% 
  make_linelist(
    id = "case_id",
    date_onset = "date_of_onset",
    gender = "gender"
  )

tags(ebola_tag)
#> $id
#> [1] "case_id"
#> 
#> $date_onset
#> [1] "date_of_onset"
#> 
#> $gender
#> [1] "gender"

# validate_linelist(ebola_tag)

# tag extra columns
ebola_tag_02 <- ebola %>% 
  make_linelist(
    id = "case_id",
    date_onset = "date_of_onset",
    gender = "gender",
    generation_number = "generation", allow_extra = TRUE
  )

tags(ebola_tag_02)
#> $id
#> [1] "case_id"
#> 
#> $date_onset
#> [1] "date_of_onset"
#> 
#> $gender
#> [1] "gender"
#> 
#> $generation_number
#> [1] "generation"

# validate rejects
validate_linelist(
  x = ebola_tag_02,
  allow_extra = TRUE,
  ref_types = tags_types(
    generation_number = c("character"), # if "integer", validation success
    allow_extra = TRUE
  )
)
#> Error: Some tags have the wrong class:
#>   - generation_number: Must inherit from class 'character', but has class 'integer'
```

<sup>Created on 2024-07-30 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Author: OK. I have added.

> 

Reviewer: before leaving the episode I think we can add a callout box called "When should I use `{linelist}`?" that redirects learners to the get started vignette. This is critical, given that `linelist` is more appropriate for ongoing/long-lasting analysis than historical/quick analysis.

Author: OK. Will add a callout.

:::::::::::::::::::: hint
```{r, eval=FALSE}
linelist::tags_types()
```
:::::::::::::::::::::::

- How many variables are there? and what are their data types?

::::::::::::::::::::::::::::::::::::::::::::::

Safeguarding is implicitly built into the linelist objects. If you try to delete any of the tagged 
columns, you will receive an error or warning message, as shown in the example below.

```{r, warning=TRUE}
new_df <- data |>
  dplyr::select(linelist::has_tag(c("id", "age")))
```

The default options for lost  tags in a linelist object is warning. However, it can be change to error message using `lost_tags_action()`. 

::::::::::::::::::::::::::::::::::::: challenge 

- Set the action for lost tags in a linelist to error as follows:

```{r, eval=FALSE}
linelist::lost_tags_action(action = "error")
```  
and re-run the above code segment.

- What do you learn from message?  

::::::::::::::::::::::::::::::::::::::::::::::::

To ensure that all tagged variables are standardized and have the correct data 
types, use the `linelist::validate_list()`, as 
shown in the example below:

```r
linelist::validate_linelist(data,
  allow_extra = FALSE
)
```

> If your dataset contains a `non-default` tag, set the argument 
> `allow_extra = TRUE` when creating the linelist object with its corresponding
> datatype. 

Reviewer: Given that tagging and validating a new tag needs using `make_linelist()`, `validate_linelist()`, and `tags_types()`. Probably better to keep it outside in the how-to guides.

I will share a PR on howto repo after this review. For the moment, we can block this sentences:

```suggestion
<!-- If your dataset requires a new tag, set the argument -->
<!-- `allow_extra = TRUE` when creating the linelist object with its corresponding-->
<!-- datatype. -->
```

> ::::::::::::::::::::::::: spoiler
> 
> ### Can we create {linelist} with `non-default` tags?
> 
> - How to add extra tags? with the `allow_extra = TRUE` argument?
> - How to validate these extra tags? by defining their specific `tags_types()`?
> 
> :::::::::::::::::::::::::

Reviewer: As mentioned in the previous comment, we can tackle those questions and provide a solution redirecting learners to the howto repo. 

Instead, I'm providing an exploratory challenge:

```suggestion
::::::::::::::::::::::::: challenge

Let's change some variable data types to identify how `linelist::validate_linelist()` will react when your input data has a variable with a different data type.

Try to:

- **Change** a variable data type,
- **Tag** the linelist, and then 
- **Validate** it

Identify the correlation between the error messages and the output of `linelist::tags_types()`.

:::::::::::::::::::::::::: hint

If we change the `age` variable from numeric to character:

- `cleaned_data %>%`
- `# change data type`
- `mutate(age_character = as.character(age)) %>%`
- `# tag`
- `linelist::make_linelist(`
- `  age = "age_character"`
- `) %>%`
- `# validate`
- `linelist::validate_linelist()`

Other frequent changes can be having a date variable like `date_onset` changed to a character, or a factor variable like `gender` changed to an integer.

::::::::::::::::::::::::::

::::::::::::::::::::::::: hint

- `# example 2`
- `cleaned_data %>% `
-   `# mess`
-   `mutate(date_onset_character = as.character(date_onset)) %>% `
-   `# tag`
-   `linelist::make_linelist(`
-     `date_onset = "date_onset_character"`
-   `) %>% `
-   `# validate`
-   `linelist::validate_linelist()`

- `# example 3`
- `cleaned_data %>% `
-   `# mess`
-   `mutate(gender = as.factor(gender)) %>% `
-   `mutate(gender_integer = as.integer(gender)) %>% `
-   `# tag`
-   `linelist::make_linelist(`
-     `gender = "gender_integer"`
-   `) %>% `
-   `# validate`
-   `linelist::validate_linelist()`

::::::::::::::::::::::::: 

:::::::::::::::::::::::::
```

A  `linelist` object resembles a data frame but offers richer features 
and functionalities. Packages that are linelist-aware can leverage these 
features. For example, you can extract a dataframe of only the tagged columns 
using the `linelist::tags_df()` function, as shown below:
```{r, warning=FALSE}
head(linelist::tags_df(data), 5)
```

> This allows, the extraction of use tagged-only columns in downstream analysis.

Reviewer: ```suggestion
This allows, the extraction of use tagged-only columns in downstream analysis, which will be useful for the next episode!
```
